pkg/urbit/jets/e/secp.c

```diff
@@ -208,3 +208,90 @@ u3we_make(u3_noun cor)
     return u3l_punt("secp-make", _cqes_make(has, prv));
   }
 }
+
+/* create a schnorr signature
+*/
+static u3_weak
+_cqes_sosi(u3_atom sk, u3_atom m, u3_atom a)
+{
+  c3_y key_y[32];
+  c3_y mes_y[32];
+  c3_y aux_y[32];
+
+  if ( (0 != u3r_bytes_fit(32, key_y, sk)) ||
+       (0 != u3r_bytes_fit(32, mes_y, m)) ||
+       (0 != u3r_bytes_fit(32, aux_y, a)) )
+  {
+    return u3m_bail(c3__exit);
+  }
+  else {
+    c3_y sig_y[64];
+
+    return
+      ( 0 == urcrypt_secp_schnorr_sign(sec_u, key_y, mes_y, aux_y, sig_y) )
+      ? u3i_bytes(64, sig_y)
+      : u3_none;
+  }
+}
+
+u3_noun
+u3we_sosi(u3_noun cor)
+{
+  u3_noun key, mes, aux;
+
+  if ( (c3n == u3r_mean(cor,
+                        u3x_sam_2,  &key,
+                        u3x_sam_6,  &mes,
+                        u3x_sam_7,  &aux,
+                        0)) ||
+       (c3n == u3ud(key)) ||
+       (c3n == u3ud(mes)) ||
+       (c3n == u3ud(aux)) )
+  {
+    return u3m_bail(c3__exit);
+  }
+  else {
+    return u3l_punt("secp-sosi", _cqes_sosi(key, mes, aux));
+  }
+}
+
+/* verify a schnorr signature
+*/
+static u3_atom
+_cqes_sove(u3_atom pk, u3_atom m, u3_atom sig)
+{
+  c3_y pub_y[32];
+  c3_y mes_y[32];
+  c3_y sig_y[64];
+
+  if ( (0 != u3r_bytes_fit(32, pub_y, pk)) ||
+       (0 != u3r_bytes_fit(32, mes_y, m)) ||
+       (0 != u3r_bytes_fit(64, sig_y, sig)) )
+  {
+    return u3m_bail(c3__exit);
+  }
+  else {
+    return __(urcrypt_secp_schnorr_veri(sec_u, sig_y, mes_y, pub_y));
```

Comment:
This return value likely includes non-deterministic results, at least allocation failure and invalid context (`sec_u`). I'm not sure if there's enough information in the secp api to distinguish deterministic verification failure from other cases. Ideally there is, and the urcrypt api can reflect it. If there is not, you'll have to "punt" (return `u3_none` and defer to nock).

Reply:
Looks like nondeterministic failures are done via `abort()` in secp256k1. This should only return deterministic results unless I'm mistaken.

Reply:
What is the `illegal_callback`? https://github.com/bitcoin-core/secp256k1/blob/c8aa516b57ab951e26a1ff5fa283d6f5612ca809/src/secp256k1.c#L34-L39

Reply:
Winds up getting assigned to [`secp256k1_default_illegal_callback_fn`](https://github.com/bitcoin-core/secp256k1/blob/c8aa516b57ab951e26a1ff5fa283d6f5612ca809/src/util.h#L29).

Reply:
And the `VERIFY_CHECK`s are compiled out by default?

Reply:
correct